Decode surrogate pairs that are split between chunks

ricea opened the pull request:
Previously, if a leading surrogate appeared at the end of a chunk it
would be translated to a replacement character. With this change, it can
form a surrogate pair with a trailing surrogate in the next chunk.

Also add tests for matched and unmatched surrogate pairs across chunks.

ricea committed:
Decode surrogate pairs that are split between chunks

Previously, if a leading surrogate appeared at the end of a chunk it
would be translated to a replacement character. With this change, it can
form a surrogate pair with a trailing surrogate in the next chunk.

Also add tests for matched and unmatched surrogate pairs across chunks.

tyoshino commented on tests/encode-utf8.html:
```diff
@@ -10,13 +10,95 @@
 const inputString = 'I \u{1F499} streams';
 const expectedOutputBytes = [73, 32, 240, 159, 146, 153, 32, 115, 116, 114, 101,
                              97, 109, 115];
+const astralCharacter = '\u{1F499}';
```
looks it's a blue heart. what does astral mean here?

ricea replied:
https://www.quora.com/Why-are-unicode-characters-outside-the-BMP-called-astral

I actually originally called the constant "plain1Character" but it was cumbersome and hard to read, so I looked for a better name.

I will add a clarifying comment.

tyoshino replied:
aha, thanks for the explanation. I didn't know that.

tyoshino replied:
how about supplementary?

tyoshino commented on tests/encode-utf8.html:
```diff
@@ -10,13 +10,95 @@
 const expectedOutputBytes = [73, 32, 240, 159, 146, 153, 32, 115, 116, 114, 101,
                              97, 109, 115];
+const astralCharacter = '\u{1F499}';
+const astralCharacterEncoded = [240, 159, 146, 153];
```
how about using hex literals here?

ricea replied:
I didn't think of that. Done.

tyoshino commented on tests/encode-utf8.html:
```diff
@@ -10,13 +10,95 @@
+const astralCharacterEncoded = [240, 159, 146, 153];
+const leading = astralCharacter[0];
+const trailing = astralCharacter[1];
+const replacementEncoded = [239, 191, 189];
```
ditto

ricea replied:
Done.

tyoshino commented on tests/encode-utf8.html:
```diff
@@ -10,13 +10,95 @@
-  assert_array_equals(array[0], expectedOutputBytes,
-                      'output should match expected bytes');
-}, 'encoding one string of UTF-8 should give one complete chunk');
+// These tests assume that the implementation correct classifies leading and
```
correct -> correctly?

ricea replied:
Done, thanks.

ricea committed:
Clarifying fixes

1. Note the meaning of astralCharacter
2. Use hexadecimal constants
3. Fix "correct classifies" -> "correctly classifies"
